### DISCUSSION TITLE
Minion: Fail early

### DISCUSSION THREAD

**Discussion by augensalat** [mojolicious/mojo#1662] [Q&A]
My application has to check a remote service that processes some long running task and eventually returns success or failure.
For that a minion task is enqueued with `attempts => 15`.

- As soon as the remote side returns "success" a `$job->finish` terminates the job on my side.
- If the remote does not deliver a final result within the 15 attempts, Minion terminates the job with a `failed` state.
- In case the remote delivers a final "failed" result within the 15 attempts, I would like to also terminate the job with a `failed` state and keep it from further attempts, but currently I do not find a way to accomplish this - calling `$job->fail` and i.e. setting `attempts = 1` does not seem possible  - the only way that seems to work is `$job->finish`, which means success.

Am I just missing the right button or is not possible?

**Comment by kraih** [comment]
You have to `retry` to change attempts and then `fail` afterwards.

**Reply by kraih** [reply]
I suppose it might be worth considering support for `$job->fail('reason', {attempts => 1})`.

**Reply by augensalat** [reply]
> You have to `retry` to change attempts and then `fail` afterwards.

But this means I had to retry once more even though I already received the final result in the current attempt, right?

**Reply by augensalat** [reply]
> I suppose it might be worth considering support for `$job->fail('reason', {attempts => 1})`.

Or, since I currently cannot imagine a use case with other parameters than this (and also setting attempts to 1 in order to stop retries requires a certain understanding about the inner processes), why not something like `$job->fail_without_retry('reason')` - or simply `$job->abort('reason')`?

**Reply by kraih** [reply]
You can retry into a named queue no worker handles. I don't want to bikeshed about the API, so i'm leaving the discussion here.

**Reply by augensalat** [reply]
> You can retry into a named queue no worker handles.

I must admit that I have no idea what you mean by that...
